src/centrality.rs

```diff
@@ -0,0 +1,217 @@
+use pyo3::prelude::*;
+
+use petgraph::graph::NodeIndex;
+use petgraph::visit::{
+    GraphBase,
+    GraphProp, // allows is_directed
+    IntoNeighborsDirected,
+    IntoNodeIdentifiers,
+    NodeCount,
+    NodeIndexable,
+};
+use std::collections::HashMap;
+
+// Correspondence to notation in Brandes 2001
+//
+// P -- predecessors
+// S -- sorted_by_distance,
+//      vertices in order of non-decreasing distance from s
+// Q -- Q
+// sigma -- sigma
+// delta -- delta
+// d -- distance
+
+pub fn betweenness_centrality<G>(
+    graph: G,
+    endpoints: bool,
+    normalized: bool,
+) -> PyResult<Vec<f64>>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + NodeCount
+        + GraphProp
+        + GraphBase<NodeId = NodeIndex>,
+    // rustfmt deletes the following comments if placed inline above
+    // + IntoNodeIdentifiers // for node_identifiers()
+    // + IntoNeighborsDirected // for neighbors()
+    // + NodeCount // for node_count
+    // + GraphProp // for is_directed
+{
+    let mut betweenness = vec![0.0; graph.node_count()];
+    for node_s in graph.node_identifiers() {
+        let is = graph.to_index(node_s);
+        let (mut sorted_by_distance, predecessors, sigma) =
+            shortest_path_for_centrality(&graph, &node_s);
+        sorted_by_distance.reverse(); // will be effectively popping from the stack
+        if endpoints {
+            _accumulate_endpoints(
+                &mut betweenness,
+                graph.node_count(),
+                &sorted_by_distance,
+                predecessors,
+                sigma,
+                is,
+            );
+        } else {
+            _accumulate_basic(
+                &mut betweenness,
+                graph.node_count(),
+                &sorted_by_distance,
+                predecessors,
+                sigma,
+                is,
+            );
+        }
+    }
+    _rescale(
+        &mut betweenness,
+        graph.node_count(),
+        normalized,
+        graph.is_directed(),
+        endpoints,
+    );
+
+    return Ok(betweenness);
+}
+
+fn _rescale(
+    betweenness: &mut Vec<f64>,
+    node_count: usize,
+    normalized: bool,
+    directed: bool,
+    endpoints: bool,
+) {
+    let mut do_scale = true;
+    let mut scale = 1.0;
+    if normalized {
+        if endpoints {
+            if node_count < 2 {
+                do_scale = false;
+            } else {
+                scale = 1.0 / (node_count * (node_count - 1)) as f64;
+            }
+        } else if node_count <= 2 {
+            do_scale = false;
+        } else {
+            scale = 1.0 / ((node_count - 1) * (node_count - 2)) as f64;
+        }
+    } else {
+        if !directed {
+            scale = 0.5;
+        } else {
+            do_scale = false;
+        }
+    }
```

Comment:
Could this be collapsed to?

```suggestion
    } else if !directed {
        scale = 0.5;
    } else {
        do_scale = false;
    }
```